File: 17th-April-Job_Sequencing_Problem/java/solution.java
/*
  This code is written by Bhaskar
*/


class Solution
{
    //Function to find the maximum profit and the number of jobs done.
    int[] JobScheduling(Job arr[], int n)
    {
        // Your code here
        int maxDeadline = 1;
        for(Job job : arr){
            maxDeadline = Math.max(maxDeadline , job.deadline);
        }
        
        Arrays.sort(arr , (a,b) -> b.profit - a.profit);
        
        int[] slots = new int[maxDeadline+1];
        Arrays.fill(slots,-1);
        
        int[] ans = {0,0} ; //ans[0]=count , ans[1] = profit
        
        for(Job j : arr){
            for(int i=j.deadline ; i>0 ; i--){
                if(slots[i]==-1){
                    slots[i]=1;
                    ans[0]++;
                    ans[1] += j.profit;
                    break;
                }
            }
        }
        return ans;
    }
}

File: 17th-April-Job_Sequencing_Problem/c++/solution.cpp
/*
  This code is written by Bhaskar
*/


class Solution 
{
    public:
    //Function to find the maximum profit and the number of jobs done.
    vector<int> JobScheduling(Job arr[], int n) 
    { 
    // Your code here
    int maxDeadline = 1;
    for(int i=0; i<n; i++){
        maxDeadline = max(maxDeadline, arr[i].dead);
    }

    sort(arr, arr+n, [](Job a, Job b){ return a.profit > b.profit; });

    int slots[maxDeadline+1];
    memset(slots, -1, sizeof(slots));

    vector<int> ans(2) ; //ans[0]=count , ans[1] = profit
    ans[0] = 0;
    ans[1] = 0;

    for(int i=0; i<n; i++){
        for(int j=arr[i].dead; j>0; j--){
            if(slots[j]==-1){
                slots[j]=1;
                ans[0]++;
                ans[1] += arr[i].profit;
                break;
            }
        }
    }
    return ans;
}
};

